### PR TITLE
Adding metrics for DNS.

### LINKED ISSUE #594: Add metrics to DNS proxy
Need to add metrics to the DNS proxy in ztunnel. Similar to what was done in Istio: https://github.com/istio/istio/blob/master/pkg/dns/client/monitoring.go.

### PR DESCRIPTION
**Pull request by nmittler**
Fixes #594.

### CODE REVIEW

File: src/proxy/dns.rs [hunk 1]
```diff
@@ -385,30 +392,86 @@ impl DnsStore {
...
         // Find the service for the requested host.
         let requested_name = Name::from(request.query().name().clone());
         let Some(service_match) = self.find_server(&client, &requested_name) else {
             // Unknown host. Forward to the upstream resolver.
-            return self.forwarder.forward(&client, request).await;
+            return self.forward(&client, request).await;
         };
 
+        // Increment counter for all requests.
+        self.metrics.increment(&DnsRequest {
```
Comment (ymesika) [inline src/proxy/dns.rs:469]: Looks like it's being incremented with similar labels either in `forward` or here.
Wondering if it should be placed once at line 455?

Comment (nmittler) [reply src/proxy/dns.rs:469]: Ah good catch. This call site should have the destination set. Updated.

**Comment by nmittler** [comment]
@ymesika 

> It seems like the tests were updated to include the metrics registry but not checking whether eventually the expected metrics were added.

I'm actually not sure where metrics are tested today. Got any links handy?

**Comment by ymesika** [comment]
> @ymesika
> 
> > It seems like the tests were updated to include the metrics registry but not checking whether eventually the expected metrics were added.
> 
> I'm actually not sure where metrics are tested today. Got any links handy?

Here for example: https://github.com/istio/ztunnel/blob/master/tests/direct.rs#L274

File: src/metrics/dns.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            requests,
+            forwarded_requests,
+            forwarded_failures,
+            forwarded_duration,
+        }
+    }
+}
+
+#[derive(Clone, Hash, Debug, PartialEq, Eq, EncodeLabelSet)]
+pub struct DnsLabels {
```
Comment (howardjohn) [inline src/metrics/dns.rs:76]: This seems like a LOT of labels. I am not sure this much is appropriate for Prometheus. Source IP is unbounded cardinality for example

Comment (howardjohn) [reply src/metrics/dns.rs:76]: I am a bit torn. I see the value but also the cost. I will think about it a bit more (and welcome other opinions here)

File: src/metrics/dns.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    request_protocol: String,
+
+    // Source workload.
+    source_workload: DefaultedUnknown<String>,
+    source_canonical_service: DefaultedUnknown<String>,
+    source_canonical_revision: DefaultedUnknown<String>,
+    source_workload_namespace: DefaultedUnknown<String>,
+    source_principal: DefaultedUnknown<Identity>,
+    source_app: DefaultedUnknown<String>,
+    source_version: DefaultedUnknown<String>,
```
Comment (howardjohn) [inline src/metrics/dns.rs:89]: Largely these are legacy I think, in favor of source_canonical_service/source_canonical_revision. I wonder if we drop them for this new metric

**Review by howardjohn — COMMENTED** [review]
I only had a chance to skim, looks reasonable. I will actually review later today

**Comment by nmittler** [comment]
@howardjohn I've removed a bunch of the metrics labels. PTAL

**Review by howardjohn — APPROVED** [review]

**Pull request merged**